turbo/stages/stageloop.go

```diff
@@ -49,67 +50,93 @@ func StageLoop(
 	sync *stagedsync.StagedSync,
 	hd *headerdownload.HeaderDownload,
 	chainConfig *params.ChainConfig,
-) error {
+) {
 	initialCycle := true
-	stopped := false
-	logEvery := time.NewTicker(logInterval)
-	defer logEvery.Stop()
 
-	for !stopped {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
 		// Estimate the current top height seen from the peer
 		height := hd.TopSeenHeight()
-
-		origin, err := stages.GetStageProgress(db, stages.Headers)
-		if err != nil {
-			return err
-		}
-		hashStateStageProgress, err1 := stages.GetStageProgress(db, stages.Bodies) // TODO: shift this when more stages are added
-		if err1 != nil {
-			return err1
+		if err := StageLoopStep(ctx, db, sync, height, chainConfig, initialCycle); err != nil {
+			log.Error("Stage loop failure", "error", err)
+			continue
 		}
 
-		canRunCycleInOneTransaction := !initialCycle && height-origin < 1024 && height-hashStateStageProgress < 1024
-
-		var writeDB ethdb.Database // on this variable will run sync cycle.
+		initialCycle = false
+	}
+}
 
-		// create empty TxDb object, it's not usable before .Begin() call which will use this object
-		// It allows inject tx object to stages now, define rollback now,
-		// but call .Begin() after hearer/body download stages
-		var tx ethdb.DbWithPendingMutations
-		if canRunCycleInOneTransaction {
-			tx, err = db.Begin(context.Background(), ethdb.RW)
-			if err != nil {
-				return err
+func StageLoopStep(
+	ctx context.Context,
+	db ethdb.Database,
+	sync *stagedsync.StagedSync,
+	highestSeenHeader uint64,
+	chainConfig *params.ChainConfig,
+	initialCycle bool,
+) (err error) {
+	// avoid crash because TG's core does many things -
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				panic(r)
 			}
-			defer tx.Rollback()
-			writeDB = tx
-		} else {
-			writeDB = db
 		}
+	}()
```

Comment:
@AlexeyAkhunov FYI: `StageLoop` will continue after error and recover panic.

Reply:
Ok